src/solver.rs

```diff
@@ -38,6 +41,161 @@ pub struct CandidateBuckets {
+    limit: Duration,  // maximum allowed elapsed time
+}
+
+impl TimeBudget {
+    /// Create a new budget that lasts for `limit` (e.g., 30 seconds).
+    fn new(limit: Duration) -> Self {
+        Self { start: Instant::now(), limit }
+    }
+
+    /// Returns true if the allowed time has fully elapsed.
+    fn expired(&self) -> bool {
+        self.start.elapsed() >= self.limit
+    }
+
+    // Returns the remaining time before expiration, or zero if the budget is already used up.
+    // Unused for now but it may be useful later
+    // fn remaining(&self) -> Duration {self.limit.saturating_sub(self.start.elapsed())}
+}
+
+
+/// Build the deterministic lookup key for a binding given the pattern's lookup vars.
+fn lookup_key_for_binding(binding: &Bindings, keys_opt: &Option<HashSet<char>>) -> LookupKey {
+    match keys_opt {
+        None => None, // pattern has no lookup constraints
+        Some(keys) => {
+            // Collect (var, value) only for vars present in this binding.
+            let mut pairs: Vec<(char, String)> = Vec::with_capacity(keys.len());
+            for &var in keys {
+                if let Some(val) = binding.get(var) {
+                    pairs.push((var, val.clone()));
+                } else {
+                    // Missing a required key → this binding is unusable for join
+                    pairs.clear();
+                    break;
+                }
+            }
+            if pairs.is_empty() && !keys.is_empty() {
+                // Some key was missing: return a key that will never match
+                // (caller will just skip pushing in this case)
+                return Some(Vec::new()); // sentinel "impossible" key
+            }
+            pairs.sort_unstable_by_key(|(c, _)| *c);
+            Some(pairs)
+        }
+    }
+}
+
+/// Push a binding into the appropriate bucket and bump the count.
+fn push_binding(words: &mut [CandidateBuckets], i: usize, key: LookupKey, binding: Bindings) {
+    words[i].buckets.entry(key).or_default().push(binding);
+    words[i].count += 1;
+}
+
+/// Scan a slice of the word list and incrementally fill candidate buckets.
+/// Returns whether we hit the per-pattern cap and the last index scanned (exclusive).
+/// Scan a slice of the word list and incrementally fill candidate buckets.
+/// Returns (new_scan_pos, time_up).
+fn scan_batch(
+    word_list: &[&str],
+    start_idx: usize,
+    batch_size: usize,
+    patterns: &Patterns,
+    parsed_forms: &[ParsedForm],
+    scan_hints: &[PatternLenHints],
+    var_constraints: &crate::constraints::VarConstraints,
+    joint_constraints: Option<&JointConstraints>,
+    words: &mut [CandidateBuckets],
+    budget: Option<&TimeBudget>,
+) -> (usize, bool) {
+    let mut i_word = start_idx;
+    let end = start_idx.saturating_add(batch_size).min(word_list.len());
+
+    while i_word < end {
+        if let Some(b) = budget {
+            if b.expired() { return (i_word, true); }
```

Comment:
we should probably have this timeout bubble all the way up so that the UI will know if there was a timeout

Reply:
so perhaps instead of returning a `(usize, bool)`, return a new type with those two components and a `bool`(?) for timedout? (for another PR, ofc)